src/xunit.performance.api/Profilers/Etw/EtwHelper.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Session;
+using Microsoft.Xunit.Performance.Api.Native.Windows;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Xunit.Performance.Api.Profilers.Etw
+{
+    /// <summary>
+    /// Provides a basic interface for ETW Tracing operations.
+    /// </summary>
+    internal static class EtwHelper
+    {
+        static EtwHelper()
+        {
+            AvailablePreciseMachineCounters = TraceEventProfileSources.GetInfo();
+            CanEnableKernelProvider = TraceEventSession.IsElevated() == true;
+        }
+
+        /// <summary>
+        /// Collection of available precise machine counters.
+        /// </summary>
+        public static IReadOnlyDictionary<string, ProfileSourceInfo> AvailablePreciseMachineCounters { get; }
+
+        /// <summary>
+        /// Indicates whether an ETW kernel session can be enabled.
+        /// </summary>
+        public static bool CanEnableKernelProvider { get; }
+
+        /// <summary>
+        /// Enable the PMC machine wide, for ETW capture.
+        /// </summary>
+        /// <param name="profileSourceInfos">Collection of PMC to be enabled.</param>
+        public static void SetPreciseMachineCounters(IReadOnlyCollection<ProfileSourceInfo> profileSourceInfos)
+        {
+            if (profileSourceInfos == null)
+                throw new ArgumentNullException($"{nameof(profileSourceInfos)} cannot be null.");
+            if (!Kernel32.IsWindows8OrGreater())
+                throw new InvalidOperationException("System Tracing is only supported on Windows 8 and above.");
+
+            var profileSourceIDs = new List<int>();
+            var profileSourceIntervals = new List<int>();
+
+            foreach (var psi in profileSourceInfos)
+            {
+                if (AvailablePreciseMachineCounters.TryGetValue(psi.Name, out var profInfo))
+                {
+                    profileSourceIDs.Add(profInfo.ID);
+                    profileSourceIntervals.Add(Math.Min(profInfo.MaxInterval, Math.Max(profInfo.MinInterval, psi.Interval)));
+                }
+            }
+
+            if (profileSourceIDs.Count > 0)
+            {
+                //
+                // FIXME: This function changes the -pmcsources intervals machine wide.
```

Comment:
This is not covered in this PR. I believe there is already an opened issue for this, but I like keeping the comment on the code too.
Hint: If you use VS Task List, this can show there.

---
In reply to: [142810222](https://github.com/Microsoft/xunit-performance/pull/241#discussion_r142810222) [](ancestors = 142810222)